File: src/main/resources/coremods/level_load.js
function initializeCoreMod() {
    var ASMAPI = Java.type('net.neoforged.coremod.api.ASMAPI');
    var Opcodes = Java.type('org.objectweb.asm.Opcodes');
    var InsnList = Java.type('org.objectweb.asm.tree.InsnList');
    var InsnNode = Java.type('org.objectweb.asm.tree.InsnNode');
    var VarInsnNode = Java.type('org.objectweb.asm.tree.VarInsnNode');
    
    return {
        'level_load': {
            'target': {
                'type': 'METHOD',
                'class': 'net.minecraft.server.level.ServerLevel',
                'methodName': '<init>',
                'methodDesc': '(Lnet/minecraft/server/MinecraftServer;Ljava/util/concurrent/Executor;Lnet/minecraft/world/level/storage/LevelStorageSource$LevelStorageAccess;Lnet/minecraft/world/level/storage/ServerLevelData;Lnet/minecraft/resources/ResourceKey;Lnet/minecraft/world/level/dimension/LevelStem;Lnet/minecraft/server/level/progress/ChunkProgressListener;ZJLjava/util/List;ZLnet/minecraft/world/RandomSequences;)V'
            },
            'transformer': function (method) {
                var target = new InsnList();
                target.add(new InsnNode(Opcodes.DUP));
                target.add(new VarInsnNode(Opcodes.ALOAD, 1));
                target.add(ASMAPI.buildMethodCall(
                    'org/moddingx/libx/impl/libxcore/CoreLevelLoad', 'startLevelLoad',
                    '(Lnet/minecraft/world/level/chunk/ChunkGenerator;Lnet/minecraft/server/MinecraftServer;)V',
                    ASMAPI.MethodType.STATIC
                ));
                for (var i = 0; i < method.instructions.size(); i++) {
                    var node = method.instructions.get(i);
                    if (node.getOpcode() == Opcodes.INVOKEVIRTUAL) {
                        var methodNode = node;
                        if (methodNode.owner == 'net/minecraft/world/level/dimension/LevelStem' && methodNode.name == 'generator' && methodNode.desc == '()Lnet/minecraft/world/level/chunk/ChunkGenerator;') {
                            method.instructions.insert(node, target);
                            return method;
                        }
                    }
                }
                throw new Error('Failed to patch ServerLevel.class');
            }
        }
    };
}

File: coremods/src/main/java/org/moddingx/libx/coremods/transformers/LevelLoad.java
package org.moddingx.libx.coremods.transformers;

import net.neoforged.neoforgespi.transformation.ProcessorName;
import net.neoforged.neoforgespi.transformation.SimpleClassProcessor;
import net.neoforged.neoforgespi.transformation.SimpleTransformationContext;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.*;

import java.util.Set;

public class LevelLoad extends SimpleClassProcessor {

    @Override
    public ProcessorName name() {
        return new ProcessorName("libx", "level_load");
    }

    @Override
    public Set<Target> targets() {
        return Set.of(new Target("net.minecraft.server.level.ServerLevel"));
    }

    @Override
    public void transform(ClassNode cn, SimpleTransformationContext ctx) {
        for (MethodNode mn : cn.methods) {
            if (!"<init>".equals(mn.name)) continue;

            for (int i = 0; i < mn.instructions.size(); i++) {
                AbstractInsnNode node = mn.instructions.get(i);
                if (node.getOpcode() == Opcodes.INVOKEVIRTUAL) {
                    MethodInsnNode methodInsnNode = (MethodInsnNode) node;
                    if (methodInsnNode.owner.equals("net/minecraft/world/level/dimension/LevelStem")
                            && methodInsnNode.name.equals("generator")
                            && methodInsnNode.desc.equals("()Lnet/minecraft/world/level/chunk/ChunkGenerator;")) {

                        InsnList injection = new InsnList();
                        injection.add(new InsnNode(Opcodes.DUP));
                        injection.add(new VarInsnNode(Opcodes.ALOAD, 1));
                        injection.add(new MethodInsnNode(
                                Opcodes.INVOKESTATIC,
                                "org/moddingx/libx/impl/libxcore/CoreLevelLoad",
                                "startLevelLoad",
                                "(Lnet/minecraft/world/level/chunk/ChunkGenerator;Lnet/minecraft/server/MinecraftServer;)V",
                                false
                        ));
                        mn.instructions.insert(node, injection);
                        return;
                    }
                }
            }
        }
        throw new IllegalStateException("Failed to patch ServerLevel.class");
    }
}
